modules/porous_flow/src/kernels/PorFlowComponentMassTimeDerivative.C

```diff
@@ -0,0 +1,95 @@
+  _dfluid_density_dvar(getMaterialProperty<std::vector<std::vector<Real> > >("dPorFlow_fluid_phase_density_dvar")),
+  _fluid_saturation(getMaterialProperty<std::vector<Real> >("PorFlow_saturation")),
+  _fluid_saturation_old(getMaterialPropertyOld<std::vector<Real> >("PorFlow_saturation")),
+  _dfluid_saturation_dvar(getMaterialProperty<std::vector<std::vector<Real> > >("dPorFlow_saturation_dvar")),
+  _mass_frac(getMaterialProperty<std::vector<std::vector<Real> > >("PorFlow_mass_frac")),
+  _mass_frac_old(getMaterialPropertyOld<std::vector<std::vector<Real> > >("PorFlow_mass_frac")),
+  _dmass_frac_dvar(getMaterialProperty<std::vector<std::vector<std::vector<Real> > > >("dPorFlow_mass_frac_dvar"))
+{
+}
+
+/// Note that this kernel lumps the mass terms to the nodes, so that there is no mass at the qp's.
+Real
+PorFlowComponentMassTimeDerivative::computeQpResidual()
+{
+  mooseAssert(_component_index < _mass_frac[_i][0].size(), "PorFlowComponentMassTimeDerivative: component_index is given as " << _component_index << " which must be less than the number of fluid components described by the mass-fraction matrix, which is " << _mass_frac[_i][0].size());
+  unsigned int num_phases = _fluid_density[_i].size();
+  mooseAssert(num_phases == _fluid_saturation[_i].size(), "PorFlowComponentMassTimeDerivative: Size of fluid density = " << num_phases << " size of fluid saturation = " << _fluid_saturation[_i].size() << " but both these must be equal to the number of phases in the system");
+
+  Real mass = 0.;
+  Real mass_old = 0.;
+  for (unsigned ph = 0; ph < num_phases; ++ph)
+  {
+    mass += _fluid_density[_i][ph]*_fluid_saturation[_i][ph]*_mass_frac[_i][ph][_component_index];
+    mass_old += _fluid_density_old[_i][ph]*_fluid_saturation_old[_i][ph]*_mass_frac_old[_i][ph][_component_index];
+   }
+
+  return _test[_i][_qp] * (_porosity[_i]*mass - _porosity_old[_i]*mass_old)/_dt;
+}
+
+Real
+PorFlowComponentMassTimeDerivative::computeQpJacobian()
+{
+  /// If the variable is not a PorFlow variable (very unusual), the diag Jacobian terms are 0
+  if (!(_var_is_porflow_var))
+    return 0.0;
+  return computeQpJac(_porflow_name_UO.porflow_var_num(_var.number()));
+}
+
+Real
+PorFlowComponentMassTimeDerivative::computeQpOffDiagJacobian(unsigned int jvar)
+{
+  /// If the variable is not a PorFlow variable, the OffDiag Jacobian terms are 0
+  if (_porflow_name_UO.not_porflow_var(jvar))
+    return 0.0;
+  return computeQpJac(_porflow_name_UO.porflow_var_num(jvar));
+}
+
+Real
+PorFlowComponentMassTimeDerivative::computeQpJac(unsigned int pvar)
+{
+  /// As mass is lumped to the nodes, only non-zero terms are for _i==_j
+  if (_i != _j)
+    return 0.0;
+
+  unsigned int num_phases = _fluid_density[_i].size();
+
+  Real dmass = 0.0;
+  for (unsigned ph = 0; ph < num_phases; ++ph)
+  {
+    dmass += _dfluid_density_dvar[_i][ph][pvar]*_fluid_saturation[_i][ph]*_mass_frac[_i][ph][_component_index]*_porosity[_i];
```

Comment:
operator spacing (also the repeated use of `+=` should probably be replaced by just letting the term flow over multiple lines and `+` operators)